google/auth/aio/transport/sessions.py

```diff
@@ -0,0 +1,252 @@
+            )
+
+        self._credentials = credentials
+
+    async def request(
+        self,
+        method: str,
+        url: str,
+        data: bytes = None,
+        headers: Mapping[str, str] = None,
+        max_allowed_time: Optional[
+            float
+        ] = None,  # TODO (ohmayr): set a default value for timeout.
+        timeout: Optional[float] = _DEFAULT_TIMEOUT,
+        **kwargs,
+    ) -> transport.Response:
+
+        """
+        Args:
+                method (str): The http method used to make the request.
+                url (str): The URI to be requested.
+                data (bytes): The payload or body in HTTP request.
+                headers (Mapping[str, str]): Request headers.       
+                timeout (Optional[float]):
+                The amount of time in seconds to wait for the server response
+                with each individual request.
+                max_allowed_time (Optional[float]):
+                If the method runs longer than this, a ``Timeout`` exception is
+                automatically raised. Unlike the ``timeout`` parameter, this
+                value applies to the total method execution time, even if
+                multiple requests are made under the hood.
+
+                Mind that it is not guaranteed that the timeout error is raised
+                at ``max_allowed_time``. It might take longer, for example, if
+                an underlying request takes a lot of time, but the request
+                itself does not timeout, e.g. if a large file is being
+                transmitted. The timout error will be raised after such
+                request completes.
+        
+        Raises:
+        # TODO (ohmayr): populate this.
+
+        Returns:
+        # TODO (ohmayr): populate this.
+
+        # TODO (ohmayr): Investigate if this is required.
+        # I think it is reasonable to agree on a strict type for headers and 
+        # let the caller handle any translation.
+            if headers:
+                for key in headers.keys():
+                    if type(headers[key]) is bytes:
+                        headers[key] = headers[key].decode("utf-8")
+
+        # Headers come in as bytes which isn't expected behavior, the resumable
+        # media libraries in some cases expect a str type for the header values,
+        # but sometimes the operations return these in bytes types.
+        """
+
+        if not isinstance(self._credentials, Credentials):
+            raise ValueError(
+                "The configured credentials are invalid and must be of type `google.auth.aio.credentials.Credentials`"
+            )
+
+        _credentials_refresh_attempt = 0
+        response = None
+
+        try:
+            async with timeout_guard(max_allowed_time) as with_timeout:
+                while (
+                    _credentials_refresh_attempt
+                    < transport.DEFAULT_MAX_REFRESH_ATTEMPTS
+                ):
```

Comment:
Can you use the `exponential_backoff` iterator instead?

Reply:
addressed.